streaming/src/main/scala/org/apache/spark/streaming/dstream/FileInputDStream.scala

```diff
@@ -173,8 +180,48 @@ class FileInputDStream[K, V, F <: NewInputFormat[K,V]](
       val filter = new PathFilter {
         def accept(path: Path): Boolean = isNewFile(path, currentTime, modTimeIgnoreThreshold)
       }
-      val newFiles = fs.listStatus(directoryPath, filter).map(_.getPath.toString)
-      val timeTaken = clock.getTimeMillis() - lastNewFileFindingTime
+      val directoryDepth = fs.getFileStatus(directoryPath).getPath.depth()
+
+      // Nested directories to find new files.
+      def dfs(status: FileStatus): List[FileStatus] = {
+        val path = status.getPath
+        val depthFilter = depth + directoryDepth - path.depth()
+        if (status.isDir) {
+          if (depthFilter - 1 >= 0) {
+            if (lastFoundDirs.contains(path)) {
+              if (status.getModificationTime > modTimeIgnoreThreshold) {
+                fs.listStatus(path).toList.flatMap(dfs(_))
+              } else Nil
+            } else {
+              lastFoundDirs += path
+              fs.listStatus(path).toList.flatMap(dfs(_))
+            }
+          } else Nil
+        } else {
+          if (filter.accept(path)) status :: Nil else Nil
+        }
+      }
+
+      val path = if (lastFoundDirs.isEmpty) Seq(fs.getFileStatus(directoryPath))
+      else {
+        lastFoundDirs.filter { path =>
+          // If the mod time of directory is more than ignore time, no new files in this directory.
+          try {
+            val status = fs.getFileStatus(path)
+            if (status != null && status.getModificationTime > modTimeIgnoreThreshold) true
```

Comment:
You can get rid of the entire `if` since you're saying `if (x) true else false`. Just write `status != null && status.getModificationTime > modTimeIgnoreThreshold`